src/FSharp.Data.GraphQL.Server.AppInfrastructure/GraphQLWebsocketMiddleware.fs

```diff
@@ -0,0 +1,375 @@
+        with
+        | :? InvalidMessageException as e ->
+          return
+            fail <| InvalidMessage(4400, e.Message.ToString())
+        | :? JsonException as e ->
+          if logger.IsEnabled(LogLevel.Debug) then
+            logger.LogDebug(e.ToString())
+          else
+            ()
+          return
+            fail <| InvalidMessage(4400, "invalid json in client message")
+    }
+
+  let isSocketOpen (theSocket : WebSocket) =
+    not (theSocket.State = WebSocketState.Aborted) &&
+    not (theSocket.State = WebSocketState.Closed) &&
+    not (theSocket.State = WebSocketState.CloseReceived)
+
+  let canCloseSocket (theSocket : WebSocket) =
+    not (theSocket.State = WebSocketState.Aborted) &&
+    not (theSocket.State = WebSocketState.Closed)
+
+  let receiveMessageViaSocket (cancellationToken : CancellationToken) (serializerOptions: JsonSerializerOptions) (executor : Executor<'Root>) (socket : WebSocket) : Task<RopResult<ClientMessage, ClientMessageProtocolFailure> option> =
+    task {
+      let buffer = Array.zeroCreate 4096
+      let completeMessage = new List<byte>()
+      let mutable segmentResponse : WebSocketReceiveResult = null
+      while (not cancellationToken.IsCancellationRequested) &&
+            socket |> isSocketOpen &&
+            ((segmentResponse = null) || (not segmentResponse.EndOfMessage)) do
+        try
+          let! r = socket.ReceiveAsync(new ArraySegment<byte>(buffer), cancellationToken)
+          segmentResponse <- r
+          completeMessage.AddRange(new ArraySegment<byte>(buffer, 0, r.Count))
+        with :? OperationCanceledException ->
+          ()
+
+      let message =
+        completeMessage
+        |> Seq.filter (fun x -> x > 0uy)
+        |> Array.ofSeq
+        |> System.Text.Encoding.UTF8.GetString
+      if String.IsNullOrWhiteSpace message then
+        return None
+      else
+        let! result =
+          message
+          |> deserializeClientMessage serializerOptions
+        return Some result
+    }
+
+  let sendMessageViaSocket (jsonSerializerOptions) (socket : WebSocket) (message : ServerMessage) =
+    task {
+      if not (socket.State = WebSocketState.Open) then
+        if logger.IsEnabled(LogLevel.Trace) then
+          logger.LogTrace("Ignoring message to be sent via socket, since its state is not 'Open', but '{state}'", socket.State)
+      else
+        let! serializedMessage = message |> serializeServerMessage jsonSerializerOptions
+        let segment =
+          new ArraySegment<byte>(
+            System.Text.Encoding.UTF8.GetBytes(serializedMessage)
+          )
+        if not (socket.State = WebSocketState.Open) then
+          if logger.IsEnabled(LogLevel.Trace) then
+            logger.LogTrace("ignoring message to be sent via socket, since its state is not 'Open', but '{state}'", socket.State)
+        else
+          do! socket.SendAsync(segment, WebSocketMessageType.Text, endOfMessage = true, cancellationToken = new CancellationToken())
```

Comment:
Shouldn't we pass some real cancellation token?
```suggestion
          do! socket.SendAsync(segment, WebSocketMessageType.Text, endOfMessage = true, cancellationToken = new CancellationToken.Default)
```

Reply:
I have to admit that I'm not so sure. I tested this code with the ChatApp sample app and, regarding the cancellation token logic, I observed that abruptly cancelling anything in lower layers just because something in an upper layer was cancelled (i.e. passing a "root" cancellation token all the way down) may cause exceptions. I'm afraid that the concept of cancellation tokens is such that the code that effectively checks the state of cancellation code does so:
* before even trying to execute its asynchronous operation;
* from time to time during the execution of its asynchronous operation

As I said, I'm unsure about how the cancellation token passed to this particular method (i.e.: `SendAsync`) can be used in a senseful way. One thing I do see as a possibility would be to pass in a timeout cancellation token, but not just forward a cancellation token we got from the outside.

Reply:
I found this blog post on the topic of cancellation tokens: [Recommended patterns for CancellationToken](https://devblogs.microsoft.com/premier-developer/recommended-patterns-for-cancellationtoken/) . I think the layer we're in in this code is past the "point of no cancellation".
When, for example, the application was requested to shut down, this shutdown request shouldn't be propagated in the cancellation token passed to `SendAsync` because this is a method that should be executed until it's done (the other end is waiting for a response, the infrastructure is still set up for communication etc). Instead, the shutdown request should be handled from the outside. In the case of this code, this is done in the main loop. If you read the code again, you'll see that the cancellation token is checked there, in the `while` condition:
```fsharp
try
  while not cancellationToken.IsCancellationRequested && socket |> isSocketOpen do
    let! receivedMessage = safe_Receive()
    match receivedMessage with
```